### DISCUSSION TITLE
Dynamic Class Name Not Applying CSS on Front-End

### DISCUSSION THREAD

**Discussion by rohit-ghoghari** [tailwindlabs/tailwindcss#14210] [Help]
// Fetching dynamic class not working in taiwind css

Right now, I am working with my project laravel as backend and next.js (App router) as frontend technology. for frontend i have use taiwind css.
I have some functionality in my project in which css need to dynamic assign from backend side and need to display in frontend side.

This is my about page:

I have TRY IT in multiple ways:

// ----------------------------------------First Option------------------------------------------------------------//

// app/about.js

/*

Issue: Dynamic Class Name Not Applying CSS on Front-End

Description:
We are using a class name that is partially dynamic. The base part of the class name, "text-", is static, while the second part, "green", is dynamic and comes from a variable ${item.class}. The expected output is "text-green", but the CSS associated with this class is not being applied on the front-end.

Example:

let colorClass = `text-${item.class}`; // where item.class could be 'green'

Expected Output:

The class "text-green" should be applied to the element, but the corresponding CSS is not working as expected.

Question:
Why is the "text-green" class not applying its CSS on the front-end, and how can we ensure that dynamically generated classes like this are correctly rendered with their associated styles?

*/

export default async function About() {
	
	//API Call to get header icon data
	 const res = await fetch('https://api.example.com/colour/data');
	 const { icons } = res.data

	 //icons data recieved

	 const icons = [
	 	{
	 		'icon' : 'Bookmarks',
	 		'class' : 'green',
	 	},
	 	{
	 		'icon' : 'FilterIcon',
	 		'class' : 'blue',
	 	},
	 	{
	 		'icon' : 'ShareIcon',
	 		'class' : 'yellow',
	 	}
	 ];

	  return (
	    <>
	      {/* Header icons */}
	      <div className="flex items-center gap-1">
	        {icons && icons.map((item, index) => {
	            return (
	              <button
	                className={`text-${item.class} bg-${item.class}`}
	              >
	              <GlobalIcon value={item.icon} />
	              </button>
	            );
	          })}
	      </div>
	    </>
	  );
};

// tailwind.config.js

const { nextui } = require("@nextui-org/react");
module.exports = {
  content: [
    "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}",
    "./node_modules/@nextui-org/theme/dist/**/*.{js,ts,jsx,tsx}",
  ],
  theme: {
    extend: {
      colors: {
        white: "#ffffff",
        green: "#189775",
        blue: "#0000FF",
        yellow: "#EFE89B",
      },

      fontFamily: {
        geist: [
          "geist"
        ],
      },
    },
  },
  plugins: [
    nextui({
      layout: {
        ariaLabel: "NextUI Layout",
      },
    }),
  ],
};

// ----------------------------------------Second Option------------------------------------------------------------//

/*

Issue: Dynamic Color code Not Applying CSS on Front-End

Description:
We are using a Color code that is partially dynamic. The base part of the color code, "text-", is static, while the second part, "#189775", is dynamic and comes from a variable ${item.class}. The expected output is "text-[#189775]", but the CSS associated with this class is not being applied on the front-end.

Example:

let colorClass = `text-[${item.class}]`; // where item.class could be 'green'

Expected Output:

The class "text-[#189775]" should be applied to the element, but the corresponding CSS is not working as expected.

Question:
Why is the "text-[#189775]" class not applying its CSS on the front-end, and how can we ensure that dynamically generated Color like this are correctly rendered with their associated styles?

*/

// app/about.js

export default async function About() {
	
	//API Call to get header icon data
	 const res = await fetch('https://api.example.com/colour/data');
	 const { icons } = res.data

	 //icons data recieved

	 const icons = [
	 	{
	 		'icon' : 'Bookmarks',
	 		'class' : '#189775',
	 	},
	 	{
	 		'icon' : 'FilterIcon',
	 		'class' : '#0000FF',
	 	},
	 	{
	 		'icon' : 'ShareIcon',
	 		'class' : '#EFE89B',
	 	}
	 ];

	  return (
	    <>
	      {/* Header icons */}
	      <div className="flex items-center gap-1">
	        {icons && icons.map((item, index) => {
	            return (
	              <button
	                className={`text-[${item.class}] bg-[${item.class}]`}
	              >
	              <GlobalIcon value={item.icon} />
	              </button>
	            );
	          })}
	      </div>
	    </>
	  );
};

// tailwind.config.js

const { nextui } = require("@nextui-org/react");
module.exports = {
  content: [
    "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}",
    "./node_modules/@nextui-org/theme/dist/**/*.{js,ts,jsx,tsx}",
  ],
  theme: {
    extend: {
      colors: {
        white: "#ffffff",
        green: "#189775",
        blue: "#0000FF",
        yellow: "#EFE89B",
      },

      fontFamily: {
        geist: [
          "geist"
        ],
      },
    },
  },
  plugins: [
    nextui({
      layout: {
        ariaLabel: "NextUI Layout",
      },
    }),
  ],
};

**Comment by wongjn** [comment]
As per [the documentation](https://tailwindcss.com/docs/content-configuration#dynamic-class-names):

> The most important implication of how Tailwind extracts class names is that it will only find classes that exist as _complete unbroken strings_ in your source files.
>
> If you use string interpolation or concatenate partial class names together, Tailwind will not find them and therefore will not generate the corresponding CSS:
>
> **Don’t construct class names dynamically**
>
> ```vue
> <div class="text-{{ error ? 'red' : 'green' }}-600"></div>
> ```
> In the example above, the strings `text-red-600` and `text-green-600` do not exist, so Tailwind will not generate those classes.
> Instead, make sure any class names you’re using exist in full:
>
> **Always use complete class names**
>
> ```vue
> <div class="{{ error ? 'text-red-600' : 'text-green-600' }}"></div>
> ``` 

So for your first situation, where you want `text-green bg-green`, etc. you could:
- Have a dictionary for `color` to Tailwind class names:
  ```jsx
  const res = await fetch('https://api.example.com/colour/data');
  const { icons } = res.data

  const COLORS = {
    green: 'text-green bg-green',
    blue: 'text-blue bg-blue',
    yellow: 'text-yellow bg-yellow',
  }

  return (
    <>
      {/* Header icons */}
      <div className="flex items-center gap-1">
        {icons && icons.map((item, index) => {
            return (
              <button
                className={ COLORS[item.class] }
  ```
 - [`safelist` the classes](https://tailwindcss.com/docs/content-configuration#safelisting-classes), if you have a limited number of known colors:
   ```js
   module.exports = {
     safelist: [
       { pattern: /^(text|bg)-(green|blue|yellow)$/ },
       // …
     ],
     // …
   ];
   ```

For when the color can be dynamic and is given as a color code, consider using the `style` attribute:
  ```jsx
  const res = await fetch('https://api.example.com/colour/data');
  const { icons } = res.data

  return (
    <>
      {/* Header icons */}
      <div className="flex items-center gap-1">
        {icons && icons.map((item, index) => {
            return (
              <button
                style={{
                  color: item.class,
                  backgroundColor: item.class
                }}
              >
  ```

**Reply by rohit-ghoghari** [reply]
Have you any solution for my Second Option issue?

**Reply by wongjn** [reply]
Yes, I mentioned it [in my original comment](https://github.com/tailwindlabs/tailwindcss/discussions/14210#discussioncomment-10366789):

> For when the color can be dynamic and is given as a color code, consider using the `style` attribute:
> ```jsx
> const res = await fetch('https://api.example.com/colour/data');
> const { icons } = res.data
> 
> return (
>   <>
>     {/* Header icons */}
>     <div className="flex items-center gap-1">
>       {icons && icons.map((item, index) => {
>           return (
>             <button
>               style={{
>                 color: item.class,
>                 backgroundColor: item.class
>               }}
>             >
> ```

**Comment by Jazz-Man** [comment]
Hello friends!

Recently I also faced the same situation when I needed to generate classes dynamically. However, in my case, my classes were not generated based on data from the database. I wanted to make it easier for myself to generate classes for the @tailwindcss/container-queries plugin.

That is, the idea was relatively simple - I have a React component in which I set a certain attribute, for example "containerQueries", in which I set an object with settings and then use a utility to generate classes on the fly. All this worked and the classes were generated, but as we all know, dynamic classes do not support tailwindcss.

I searched for a solution for a long time, didn't find it, and even decided to give up on the idea the next day. But then I decided to experiment with [transform](https://tailwindcss.com/docs/content-configuration#transforming-source-files). As it turned out from the documentation, roughly speaking, tailwindcss searches for classes using regex, no matter where they are in the content.

In short, I have a solution for you, perhaps not perfect, but at least working.

> It is very important that all this works in your configuration for prettier to have the option 
> ```
> “trailingComma": “none”,
> ```

First, I put all the values for the containers into a separate constant (of course, this constant is then added to your config according to the doc for @tailwindcss/container-queries):

**containers.ts:**
``` typescript
export const containers = {
	xs: '20rem',
	sm: '24rem',
	md: '28rem',
	lg: '32rem',
	xl: '36rem',
	'2xl': '42rem',
	'3xl': '48rem',
	'4xl': '56rem',
	'5xl': '64rem',
	'6xl': '72rem',
	'7xl': '80rem',
};

export type TContainerQueries = keyof typeof containers;

```

Just to understand these examples - my entire configuration is placed in a separate package and there is no point in showing it all here. So, 

**type.ts:**
``` typescript

import type { ClassValue } from 'tailwind-variants';
import type { TContainerQueries as TQueries } from '@utils/tailwind-config/src/containers';

export type TContainerQueries = {
	name?: string;
	queries: Partial<Record<TQueries, ClassValue>>;
};

```

**tailwind.config.ts:**
``` typescript

import type { ContentConfig } from 'tailwindcss/types/config';

import tvConfig from '@utils/tailwind-config';

import nextuiComponents from './node_modules/@nextui-org/react/dist/components.json';

import { containers } from '@utils/tailwind-config/src/containers';
import { TContainerQueries } from '@web/ui/types';
import { dependencies } from './package.json';

const appDependencies = new Map<string, string>(Object.entries(dependencies));

const uiComponents = nextuiComponents?.map((component) => ({
	name: component.name,
	package: component.package
}));

const contentDirs: string[] = ['pages', 'components', 'layout', 'app', 'src'];

const files: string[] = [
	`./{${contentDirs.join(',')}}/**/*.{ts,tsx}`,
	'./node_modules/@web/ui/src/*.{ts,tsx}'
];

if (uiComponents.length > 0) {
	uiComponents.map((component) => {
		if (appDependencies.has(component.package)) {
			files.push(
				`./node_modules/@nextui-org/theme/dist/components/${component.name}.js`
			);
		}
	});
}

type TRegExpExecArray = Omit<RegExpExecArray, 'groups'> & {
	groups?: {
		queries?: string;
	};
};

/**
 * Generate the correct JSON string
 * @param content
 */

const fixJsonString = (content: string) =>
	content
		.replace('{{', '{')
		.replace('}}', '}')
		.replace(/'/g, '"')
		.replace(/[\n|\t]/g, '')
		.replace(/([{,])(\w+):/gm, '$1"$2":')
		.trim()
		.toString();

const config: ContentConfig = {
	files,
	transform: {
		tsx: (content) => {
			// Pattern for finding the containerQueries attribute in the component code
			const pattern =
				/<\w+\s[^<>]*containerQueries=(?<queries>:{{([^{}]*)}}|{{([^"]*)}})(?=[^<>]*>)/gm;

			if (!pattern.test(content)) {
				return content;
			}

			let classComments = '';

			// We support the uniqueness of css classes
			const containerClassesSet = new Set<string>();

			let res: TRegExpExecArray | null;

			// Here we go through all the content and look for all the matches
			while ((res = pattern.exec(content)) !== null) {
				// This is necessary to avoid infinite loops with zero-width matches
				if (res.index === pattern.lastIndex) {
					pattern.lastIndex++;
				}

				if (!res.groups?.queries) {
					continue;
				}

				try {
					const data = fixJsonString(res.groups.queries);

					const containerQueries = JSON.parse(
						data
					) as TContainerQueries;

					for (const property in containers) {
						if (
							!containerQueries?.queries?.hasOwnProperty(property)
						) {
							continue;
						}

						const containerName = [property, containerQueries.name]
							.filter(Boolean)
							.join('/');

						// @ts-ignore
						const value = containerQueries?.queries?.[
							property
						] as string;

						const classes = value
							.replace(/\s+/g, ' ')
							.trim()
							.split(' ');

						classes.map((string: string) => {
							containerClassesSet.add(
								`@${containerName}:${string}`
							);
						});
					}
				} catch (e) {
					console.error(e);
				}
			}

			if (containerClassesSet.size > 0) {
				// Generate a valid comment in the jsx code and add it to the very beginning of the file, also add a few “\n”
				const classes = [...containerClassesSet].join(' ');

				classComments = `// ${classes}\n\n`;
			}

			return `${classComments}${content}`;
		}
	}
};

export default tvConfig(config);

```

Here is an example of a component:

**Heading.tsx:**
``` typescript
import { forwardRef } from '@nextui-org/system-rsc';
import { tv } from '@nextui-org/theme';

import type { VariantProps } from 'tailwind-variants';
import type { TBaseUiProps } from '../types';
import { useComponentProps } from '../utils';
import { boxVariants } from './box';

export type THeadingTag = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span';

export const headingVariants = tv(
	{
		extend: boxVariants,
		base: 'heading uppercase',
		variants: {
			variant: {
				archivePageTitle: 'text-42 md:text-80 xl:text-100',
				bannerTitle: 'text-44 md:text-90',
				creatorsTitle:
					'text-54 md:text-100 tracking-[-2.7px] md:tracking-[-5px]',
				roadMapTitle: 'text-100 md:text-176'
			}
		},
		defaultVariants: {
			font: 'rockstar',
			weight: 'bold',
			lineHeight: 'base'
		}
	},
	{
		responsiveVariants: true
	}
);

type TVariant = VariantProps<typeof headingVariants>;

export type THeadingProps = TBaseUiProps<THeadingTag> & Omit<TVariant, 'theme'>;

const Heading = forwardRef<'h1', THeadingProps>(
	({ displayName = 'heading', ...props }, ref) => {
		const { Component, getProps } = useComponentProps<THeadingProps>(
			{ ...props, displayName },
			headingVariants,
			'h1'
		);

		return <Component ref={ref} {...getProps()} />;
	}
);
Heading.displayName = 'Heading';

export { Heading };
```

**useComponentProps.tsx:**
``` typescript
import { clsx, objectToDeps } from '@nextui-org/shared-utils';
import type { As, PropGetter } from '@nextui-org/system';
import { mapPropsVariants } from '@nextui-org/system-rsc';
import { containers } from '@utils/tailwind-config/src/containers';
import { useMemo } from 'react';
import { cnBase } from 'tailwind-variants';
import type { TBaseUiProps, TtvReturn } from '../types';
import { inDevEnvironment, isEmptyObject } from './index';

type Props = Omit<TBaseUiProps, 'ref'>;

const parseClassesToArray = (
	string: string,
	baseClasses: string | undefined
): string[] =>
	String(string)
		.replace(baseClasses ? baseClasses : '', '')
		.replace(/\s+/g, ' ')
		.trim()
		.split(' ');

export function useComponentProps<
	T extends Record<string, any> = Props,
	V extends TtvReturn = any
>(originalProps: T, variants: V, Tag: As = 'div') {
	const [props, variantProps] = mapPropsVariants(
		originalProps,
		variants.variantKeys
	);

	const { as, className, displayName, containerQueries, ...otherProps } =
		props as Props;
	const Component = as || Tag;

	const { styles, dataAttr, containerClasses } = useMemo(
		function () {
			const dataAttr: Record<string, string | undefined> = {};

			const containerClasses: string[] = [];

			if (inDevEnvironment) {
				/**
				 * Add the variation name as a css class to improve debugging
				 */

				for (const [key, value] of Object.entries(variantProps)) {
					dataAttr[`data-variant-${key.toLowerCase()}`] = value;
				}

				dataAttr['data-ui-component'] = displayName;
			}

			if (!isEmptyObject(containerQueries)) {
				const base = variants.base ? cnBase(variants.base) : '';

				for (const property in containers) {
					if (containerQueries?.queries?.hasOwnProperty(property)) {
						const containerName = [property, containerQueries.name]
							.filter(Boolean)
							.join('/');

						// @ts-ignore
						const value = containerQueries?.queries?.[property];

						containerClasses.push(
							...parseClassesToArray(value, base).map(
								(v: string) => `@${containerName}:${v}`
							)
						);
					}
				}
			}

			const styles = variants({
				...variantProps,
				className
			});

			return { styles, dataAttr, containerClasses };
		},
		// eslint-disable-next-line react-hooks/exhaustive-deps
		[
			className,
			objectToDeps(variantProps),
			objectToDeps(containerQueries),
			variants,
			displayName
		]
	);

	const getProps: PropGetter = (props = {}) => ({
		...props,
		...otherProps,
		...dataAttr,
		className: clsx(styles, props.className, containerClasses),
		style: {
			...props.style,
			...otherProps.style
		}
	});

	return { Component, getProps };
}

```

### And most importantly, an example of using the component:

``` typescript
import { Box, Heading } from '@web/ui';

export const MyComponent = () => {
	return (
		<Box className="@container/amb">
			<Heading
				className="text-u-13xl top-[var(--u-grid-row-size)] z-0 h-full"
				containerQueries={{
					name: 'amb',
					queries: {
						xs: 'some-test-class test',
						'3xl': 'absolute'
					}
				}}
				fontFamily="rockstar"
				fontWeight="700"
				lineHeight={1}
			>
				Our ambassadors
			</Heading>
		</Box>
	);
};
```

**Comment by LinThitHtwe** [comment]
To my understanding, Tailwind CSS does not support dynamic classes. You can read more about it in the [Tailwind documentation](https://tailwindcss.com/docs/content-configuration).

Here’s a quote from the documentation:

> The most important implication of how Tailwind extracts class names is that it will only find classes that exist as complete unbroken strings in your source files.
>
> If you use string interpolation or concatenate partial class names together, Tailwind will not find them and therefore will not generate the corresponding CSS.

What you could try is, instead of using something like `text-{item.class}`, you need to provide the entire class name. 

```jsx
function Button({ color, children }) {
  const colorVariants = {
    blue: 'bg-blue-600 hover:bg-blue-500',
    red: 'bg-red-600 hover:bg-red-500',
  };

  return (
    <button className={`${colorVariants[color]} ...`}>
      {children}
    </button>
  );
} 
```

You can look up other examples in the [documentation](https://tailwindcss.com/docs/content-configuration).

Another way to overcome this is to apply styles directly instead of class names:
```jsx
const dynamicColors = {
  first: {
    bgColor: "rgba(132, 97, 241, 0.92)",
    fontColor: "#f8f8f8",
  },
  second: {
    bgColor: "#ff00ff",
    fontColor: "#f8f8f8",
  },
};

const item = item?.item; // Example purpose
const colors = dynamicColors[item as keyof typeof dynamicColors];

return (
  <span
    style={{
      backgroundColor: colors.bgColor,
      color: colors.fontColor,
    }}
    className="normal tailwind class"
  >
    Some text...
  </span>
);
 ```


Also, here's a [link](https://stackoverflow.com/questions/69687530/dynamically-build-classnames-in-tailwindcss) to a StackOverflow post discussing this topic.

I hope this help you out.
